src/clp_ffi_js/ir/StreamReader.cpp

```diff
@@ -48,204 +33,40 @@ auto StreamReader::create(DataArrayTsType const& data_array) -> StreamReader {
-        -> DecodedResultsTsType {
-    if (use_filter && false == m_filtered_log_event_map.has_value()) {
-        return DecodedResultsTsType{emscripten::val::null()};
-    }
-
-    size_t length{0};
-    if (use_filter) {
-        length = m_filtered_log_event_map->size();
-    } else {
-        length = m_encoded_log_events.size();
-    }
-    if (length < end_idx || begin_idx > end_idx) {
-        return DecodedResultsTsType{emscripten::val::null()};
-    }
-
-    std::string message;
-    constexpr size_t cDefaultReservedMessageLength{512};
-    message.reserve(cDefaultReservedMessageLength);
-    auto const results{emscripten::val::array()};
-
-    for (size_t i = begin_idx; i < end_idx; ++i) {
-        size_t log_event_idx{0};
-        if (use_filter) {
-            log_event_idx = m_filtered_log_event_map->at(i);
-        } else {
-            log_event_idx = i;
-        }
-        auto const& log_event{m_encoded_log_events[log_event_idx]};
-
-        auto const parsed{log_event.get_message().decode_and_unparse()};
-        if (false == parsed.has_value()) {
-            SPDLOG_ERROR("Failed to decode message.");
-            break;
-        }
-        message = parsed.value();
 
-        m_ts_pattern.insert_formatted_timestamp(log_event.get_timestamp(), message);
-
-        EM_ASM(
-                { Emval.toValue($0).push([UTF8ToString($1), $2, $3, $4]); },
-                results.as_handle(),
-                message.c_str(),
-                log_event.get_timestamp(),
-                log_event.get_log_level(),
-                log_event_idx + 1
+        return std::unique_ptr<IrStreamReader>(
+                new IrStreamReader(std::move(stream_reader_data_context))
         );
     }
+    SPDLOG_CRITICAL("Unable to create reader for CLP stream with version {}.", version);
 
-    return DecodedResultsTsType(results);
-}
-
-StreamReader::StreamReader(
-        StreamReaderDataContext<four_byte_encoded_variable_t>&& stream_reader_data_context
-)
-        : m_stream_reader_data_context{std::make_unique<
-                  StreamReaderDataContext<four_byte_encoded_variable_t>>(
-                  std::move(stream_reader_data_context)
-          )},
-          m_ts_pattern{m_stream_reader_data_context->get_deserializer().get_timestamp_pattern()} {}
-
-auto StreamReader::create_data_context(
-        std::unique_ptr<clp::streaming_compression::zstd::Decompressor>&& zstd_decompressor,
-        clp::Array<char> data_buffer
-) -> StreamReaderDataContext<four_byte_encoded_variable_t> {
-    rewind_reader_and_validate_encoding_type(*zstd_decompressor);
-
-    auto result{
-            clp::ir::LogEventDeserializer<four_byte_encoded_variable_t>::create(*zstd_decompressor)
+    throw ClpFfiJsException{
+            clp::ErrorCode::ErrorCode_Unsupported,
+            __FILENAME__,
+            __LINE__,
+            std::format("Unable to create reader for CLP stream with version {}.", version)
```

Comment:
```suggestion
            std::format("Unable to create reader for IR stream with version {}.", version)
```

Reply:
> junhao prefers no IR here

Sorry! Pretty sure it was me who made a suggestion in this line but the suggestion was wrong. Using `IR` would be more clear in this context since the stream reader is IR specific.

Just so we don't confuse ourselves and to make sure we're on the same page, there're cases we want to avoid certain terms:
1. IRv1 / IRv2: instead UnstructuredIR / KvPairsIR should be used
2. StreamReaderDataContext: I was trying to avoid mentioning IR at all since the class once templated with deserializer type, it can be used for Single Archive CLP as well. (That said, we have a pending discussion at https://github.com/y-scope/clp-ffi-js/pull/26#discussion_r1830665669

Reply:
1. How about UnstructuredIr and StructuredIr? That's what we were thinking in https://github.com/y-scope/clp-ffi-js/pull/26#discussion_r1830241353 I know it's a departure from what we're using in clp-core, but clp-core's naming convention is also confusing by itself since the UnstructuredIr format doesn't have a special name. And hopefully once we support backwards compatibility in the new IR format, we won't have any of these issues any more.

Reply:
Sounds good